src/volume_calc.cpp

```diff
@@ -90,182 +116,234 @@ std::vector<VolumeCalculation::Result> VolumeCalculation::execute() const
+          total_hits += master_hits[i_domain][j];
+          double f = static_cast<double>(master_hits[i_domain][j]) / total_samples;
+          double var_f = f*(1.0 - f) / total_samples;
+
+          int i_material = master_indices[i_domain][j];
+          if (i_material == MATERIAL_VOID) continue;
+
+          const auto& mat = model::materials[i_material];
+          for (int k = 0; k < mat->nuclide_.size(); ++k) {
+            // Accumulate nuclide density
+            int i_nuclide = mat->nuclide_[k];
+            atoms(i_nuclide, 0) += mat->atom_density_[k] * f;
+            atoms(i_nuclide, 1) += std::pow(mat->atom_density_[k], 2) * var_f;
+          }
+        }
+
+        // Determine volume
+        result.volume[0] = static_cast<double>(total_hits) / total_samples * volume_sample;
+        result.volume[1] = std::sqrt(result.volume[0]
+          * (volume_sample - result.volume[0]) / total_samples);
+        result.iterations = iterations;
+
+        // update threshold value if needed
+        if (trigger_type_ != TriggerMetric::not_active) {
+          double val = 0.0;
+          switch (trigger_type_) {
+            case TriggerMetric::standard_deviation:
+              val = result.volume[1];
+              break;
+            case TriggerMetric::relative_error:
+              val = result.volume[0] == 0.0 ? INFTY : result.volume[1] / result.volume[0];
+              break;
+            case TriggerMetric::variance:
+              val = result.volume[1] * result.volume[1];
               break;
-            }
+          }
+          // update max if entry is valid
+          if (val > 0.0) { trigger_val = std::max(trigger_val, val); }
+        }
+
+        for (int j = 0; j < n_nuc; ++j) {
+          // Determine total number of atoms. At this point, we have values in
+          // atoms/b-cm. To get to atoms we multiply by 10^24 V.
+          double mean = 1.0e24 * volume_sample * atoms(j, 0);
+          double stdev = 1.0e24 * volume_sample * std::sqrt(atoms(j, 1));
+
+          // Convert full arrays to vectors
+          if (mean > 0.0) {
+            result.nuclides.push_back(j);
+            result.atoms.push_back(mean);
+            result.uncertainty.push_back(stdev);
           }
         }
       }
-    } else {
-      int q = master_indices[i_domain].size();
-      int buffer[2*q];
-      for (int k = 0; k < q; ++k) {
-        buffer[2*k] = master_indices[i_domain][k];
-        buffer[2*k + 1] = master_hits[i_domain][k];
-      }
+    } // end domain loop
+
+    // if no trigger is applied, we're done
+    if (trigger_type_ == TriggerMetric::not_active) { return results; }
 
-      MPI_Send(&q, 1, MPI_INTEGER, 0, 0, mpi::intracomm);
-      MPI_Send(&buffer[0], 2*q, MPI_INTEGER, 0, 1, mpi::intracomm);
+#ifdef OPENMC_MPI
+    // update maximum error value on all processes
+    if (mpi::master) {
+      MPI_Bcast(&trigger_val, 1, MPI_DOUBLE, mpi:intracomm);
```

Comment:
Missing the `root` argument, and you have a typo too
```suggestion
      MPI_Bcast(&trigger_val, 1, MPI_DOUBLE, 0, mpi::intracomm);
```